### PR TITLE
GH-126833: Dumps graphviz representation of executor graph.

### PR DESCRIPTION
**Pull request by markshannon**
Dumps out the executor graph in graphviz format.
I'm not sure where to the put the function, for now I've named it `sys._dump_tracelets()`

It definitely shouldn't be part of the public API.

* Issue: gh-126833

### CODE REVIEW

File: Python/optimizer.c [hunk 1]
```diff
@@ -1,6 +1,7 @@
+#include "Python.h"
+
 #ifdef _Py_TIER2
 
-#include "Python.h"
```
Comment (Fidget-Spinner) [inline Python/optimizer.c:3]: Why this change?

File: Python/optimizer.c [hunk 2]
```diff
@@ -1709,4 +1716,111 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
     _Py_Executors_InvalidateAll(interp, 0);
 }
 
+static void write_str(PyObject *str, FILE *out)
```
Comment (Fidget-Spinner) [inline Python/optimizer.c]:
```suggestion
static void
write_str(PyObject *str, FILE *out)
```

File: Python/optimizer.c [hunk 3]
```diff
@@ -1709,4 +1716,111 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+        PyErr_Clear();
+        return;
+    }
+    const char *encoded_str = PyBytes_AsString(encoded_obj);
+    Py_ssize_t encoded_size = PyBytes_Size(encoded_obj);
+    fwrite(encoded_str, 1, encoded_size, out);
+    Py_DECREF(encoded_obj);
+}
+
+static int find_line_number(PyCodeObject *code, _PyExecutorObject *executor)
```
Comment (Fidget-Spinner) [inline Python/optimizer.c]:
```suggestion
static int
find_line_number(PyCodeObject *code, _PyExecutorObject *executor)
```

**Review by Fidget-Spinner — COMMENTED** [review]
Some PEP 7 nits.

At risk of being pedantic, can the graph arrow be based on the direction of the branch taken? That is, when the branch is false it points left, and when true it goes right. This way we can tell the directionality of the branch without having to refer to the source code.

**Comment by markshannon** [comment]
> can the graph arrow be based on the direction of the branch taken?

There is no direction to the arrow. It is up to the layout tool to decide where to put things.

File: Python/optimizer.c (hunk 1 above)
Comment (markshannon) [reply Python/optimizer.c:3]: Because we need the header for the `#else` clause

File: Modules/_testinternalcapi.c [hunk 4]
```diff
@@ -2081,7 +2081,6 @@ get_tracked_heap_size(PyObject *self, PyObject *Py_UNUSED(ignored))
 {
     return PyLong_FromInt64(PyInterpreterState_Get()->gc.heap_size);
 }
-
```
Comment (ericsnowcurrently) [inline Modules/_testinternalcapi.c]: I'm guessing this was unintentional. 😄

File: Python/optimizer.c [hunk 5]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+#ifdef Py_STATS
+        fprintf(out, "        <tr><td port=\"i%d\" border=\"1\" >%s -- %" PRIu64 "</td></tr>\n", i, opname, inst->execution_count);
+#else
+        fprintf(out, "        <tr><td port=\"i%d\" border=\"1\" >%s</td></tr>\n", i, opname);
+#endif
+        if (inst->opcode == _EXIT_TRACE || inst->opcode == _JUMP_TO_TOP) {
+            break;
+        }
+    }
+    fprintf(out, "    label = </table>>\n");
```
Comment (ericsnowcurrently) [inline Python/optimizer.c]: I'm sure it's fine but for my own instruction: why is `label = ` needed here but not for all the rows?  (My something-feels-off meter was twitching.)

File: Python/optimizer.c [hunk 6]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+        i += _PyOpcode_Caches[_Py_GetBaseCodeUnit(code, i).op.code];
+    }
+    return -1;
+}
+
+static void
+dump_executor(_PyExecutorObject *executor, FILE *out)
+{
+    PyCodeObject *code = executor->vm_data.code;
+    fprintf(out, "executor_%p [\n", executor);
```
Comment (ericsnowcurrently) [inline Python/optimizer.c:1763]: A few basic comments indicating what's going on--at a high level--would be helpful.

For instance (assuming I've understood correctly):

```suggestion

    /* Add a graph node for the executor. */
    fprintf(out, "executor_%p [\n", executor);
```

File: Python/optimizer.c [hunk 7]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+    fprintf(out, "    shape = none\n");
+    fprintf(out, "    label = <<table border=\"0\" cellspacing=\"0\">\n");
```
Comment (ericsnowcurrently) [inline Python/optimizer.c:1767]:
```suggestion

    /* The body of the node is an HTML table with info about the executor.
       Any table row with a "port" attribute may be used later as the endpoint
       of a graph edge, identified by "executor_<id>:<port>".
       The first row has the "start" port, */
    fprintf(out, "    label = <<table border=\"0\" cellspacing=\"0\">\n");
```

File: Python/optimizer.c [hunk 8]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+    if (code == NULL) {
+        fprintf(out, "        <tr><td border=\"1\" >No code object</td></tr>\n");
+    }
+    else {
+        fprintf(out, "        <tr><td  border=\"1\" >");
+        write_str(code->co_qualname, out);
+        int line = find_line_number(code, executor);
+        fprintf(out, ": %d</td></tr>\n", line);
+    }
+    for (uint32_t i = 0; i < executor->code_size; i++) {
```
Comment (ericsnowcurrently) [inline Python/optimizer.c:1778]:
```suggestion

    /* Add a row to the table for each each instruction in the executor's trace.
       Each one has a unique port, "i<trace index>", which means it can be used
       as the endpoint of a graph edge. */
    for (uint32_t i = 0; i < executor->code_size; i++) {
```

File: Python/optimizer.c [hunk 9]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+#else
+        fprintf(out, "        <tr><td port=\"i%d\" border=\"1\" >%s</td></tr>\n", i, opname);
+#endif
+        if (inst->opcode == _EXIT_TRACE || inst->opcode == _JUMP_TO_TOP) {
+            break;
+        }
+    }
+    fprintf(out, "    label = </table>>\n");
+    fprintf(out, "]\n\n");
+    for (uint32_t i = 0; i < executor->code_size; i++) {
```
Comment (ericsnowcurrently) [inline Python/optimizer.c:1800]:
```suggestion

    /* For each instruction in the trace that exits to an executor,
       add an edge between the instruction (table row) and the target executor. */
    for (uint32_t i = 0; i < executor->code_size; i++) {
```

File: Python/optimizer.c [hunk 10]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+        int opcode = instr->op.code;
+        if (opcode == ENTER_EXECUTOR) {
+            _PyExecutorObject *exec = code->co_executors->executors[instr->op.arg];
+            if (exec == executor) {
+                return PyCode_Addr2Line(code, i*2);
+            }
+        }
+        i += _PyOpcode_Caches[_Py_GetBaseCodeUnit(code, i).op.code];
+    }
+    return -1;
```
Comment (ericsnowcurrently) [inline Python/optimizer.c:1749]: Line numbers are 1-based, right?  We could return `0` here.  For me, `-1` indicates an error case.  I suppose lno-not-found is kind of an error case (with no exception set).  That said, don't we usually use a line number of `0` to indicate this sort of thing?

Regardless, it probably doesn't matter much either way. 😄

File: Python/optimizer.c [hunk 11]
```diff
@@ -1709,4 +1716,113 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+    fprintf(out, "    label = <<table border=\"0\" cellspacing=\"0\">\n");
+    fprintf(out, "        <tr><td port=\"start\" border=\"1\" ><b>Executor</b></td></tr>\n");
+    if (code == NULL) {
+        fprintf(out, "        <tr><td border=\"1\" >No code object</td></tr>\n");
+    }
+    else {
+        fprintf(out, "        <tr><td  border=\"1\" >");
+        write_str(code->co_qualname, out);
+        int line = find_line_number(code, executor);
+        fprintf(out, ": %d</td></tr>\n", line);
```
Comment (ericsnowcurrently) [inline Python/optimizer.c:1776]: Would it be worth handling the lno-not-found case?

For example:

```suggestion
        int line = find_line_number(code, executor);
        if (line > 0) {  /* lno is 1-based. */
            fprintf(out, ": %d", line);
        }
        fprintf(out, "</td></tr>\n", line);
```

That said, what you have is probably fine.  I doubt it would be confusing.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
This mostly looks okay.  I'm not super familiar with graphviz, but otherwise the logic makes sense.  There's just two minor comment about the return value of `find_line_number()`.  Other than that, the other comments are essentially just suggestions for comments clarifying what's going on.  As a reader, it took me a minute to sort things out, since I don't really know much graphviz, and the suggested comments would have helped.

Also, what are the options for testing this?

File: Python/optimizer.c (hunk 5 above)
Comment (markshannon) [reply Python/optimizer.c]: Indeed, the "label =" is not needed here.

File: Python/optimizer.c (hunk 7 above)
Comment (markshannon) [reply Python/optimizer.c:1767]: If we explain ports, why not cellspacing, border, etc?
I've added a link to the graphviz docs and examples, as that explains it better than I could.

File: Python/optimizer.c (hunk 10 above)
Comment (markshannon) [reply Python/optimizer.c:1749]: We generally represent no line number as -1 internally (and as `None` in Python).

We could print "None" instead of -1, but I think it best to show the internal representation as this tool is mostly for internal use and showing how the JIT works.

File: Modules/_testinternalcapi.c (hunk 4 above)
Comment (markshannon) [reply Modules/_testinternalcapi.c]: Yes

File: Python/optimizer.c (hunk 7 above)
Comment (ericsnowcurrently) [reply Python/optimizer.c:1767]: Ports are the only part in your PR that is relevant to relating nodes to edges.  Plus, ports are an unfamiliar concept while the others are not.  All that said, I suppose it isn't too hard to figure out if you think a comment isn't worth it.

File: Python/optimizer.c [hunk 12]
```diff
@@ -1709,4 +1716,131 @@ _Py_Executors_InvalidateCold(PyInterpreterState *interp)
...
+         * be placed correctly. If a row is marked `port=17`,
+         * then the outgoing edge is `{EXEC_NAME}:17 -> {TARGET}`
+         * https://graphviz.readthedocs.io/en/stable/manual.html#node-ports-compass
+         */
+        _PyUOpInstruction const *inst = &executor->trace[i];
+        const char *opname = _PyOpcode_uop_name[inst->opcode];
+#ifdef Py_STATS
+        fprintf(out, "        <tr><td port=\"i%d\" border=\"1\" >%s -- %" PRIu64 "</td></tr>\n", i, opname, inst->execution_count);
+#else
+        fprintf(out, "        <tr><td port=\"i%d\" border=\"1\" >%s</td></tr>\n", i, opname);
```
Comment (Fidget-Spinner) [inline Python/optimizer.c:1790]: Can you please add oparg and operand0 information as well? I've found it incredibly useful to debug polymorphism.

Comment (markshannon) [reply Python/optimizer.c:1790]: Not in this PR.
Happy to do so in another PR, or review yours 🙂

**Review by Fidget-Spinner — APPROVED** [review]
